src/cloud-api-adaptor/pkg/podnetwork/common.go

```diff
@@ -4,42 +4,270 @@
+func findPrimaryInterface(ns netops.Namespace) (string, netip.Addr, error) {
+	var primaryDev string
+	var primaryPrefix string
+	var primaryNum = math.MaxInt
+	var priority = math.MaxInt
+	var gw netip.Addr
 
 	routes, err := ns.RouteList(&netops.Route{Destination: netops.DefaultPrefix})
 	if err != nil {
-		return "", fmt.Errorf("failed to get routes on namespace %q: %w", ns.Path(), err)
+		return "", gw, fmt.Errorf("failed to get routes on namespace %q: %w", ns.Path(), err)
 	}
 
-	var priority = math.MaxInt
-	var dev string
-
 	for _, r := range routes {
+		// Default route check
 		if r.Destination.Bits() == 0 && r.Priority < priority {
-			dev = r.Device
+			dev := r.Device
+			devPrefix, devNum := extractInterfaceNumber(dev)
+
+			// If we haven't selected any device yet, or:
+			// The prefix matches the current primary one and has a lower number
+			// We expect the routes to be on same interface types on the pod VM
+			// eg. ens0/ens1 or eth0/eth1 etc.
+			if primaryDev == "" || (devPrefix == primaryPrefix && devNum < primaryNum) {
+				primaryDev = dev
+				primaryPrefix = devPrefix
+				primaryNum = devNum
+				gw = r.Gateway
+			}
 		}
 	}
 
-	if dev == "" {
-		return "", fmt.Errorf("failed to identify destination interface of default gateway on network namespace %q", ns.Path())
+	if primaryDev == "" {
+		return "", gw, fmt.Errorf("failed to identify primary interface on network namespace %q", ns.Path())
+	}
+
+	logger.Printf("Primary interface: %s, Gateway address: %s", primaryDev, gw.String())
+
+	return primaryDev, gw, nil
+}
+
+func setupExternalNetwork(hostNS netops.Namespace, hostPrimaryInterface string, podNS netops.Namespace) error {
+
+	// Get the secondary interface details
+	secIface, secAddrCIDR, secRoute, err := getSecondaryInterfaceDetails(hostNS, hostPrimaryInterface)
+	if err != nil {
+		return fmt.Errorf("failed to get secondary interface details: %w", err)
+	}
+
+	// Move the secondary interface to the pod network namespace
+	err = moveInterfaceToNamespace(hostNS, podNS, secIface, secAddrCIDR, secRoute)
+	if err != nil {
+		return err
+	}
+
+	return nil
+
+}
+
+func getInterfaceDetails(ns netops.Namespace, iface string) (
+	addrCIDR netip.Prefix, defRoute *netops.Route, err error) {
+
+	link, err := ns.LinkFind(iface)
+	if err != nil {
+		return netip.Prefix{}, nil, fmt.Errorf("failed to find link %q: %w", iface, err)
+	}
+
+	addrs, err := link.GetAddr()
+	if err != nil {
+		return netip.Prefix{}, nil, fmt.Errorf("failed to get addresses for link %q: %w", link.Name(), err)
+	}
+
+	for _, a := range addrs {
+		if a.Addr().Is4() {
+			addrCIDR = a
+			break
+		}
+	}
+
+	routes, err := ns.RouteList(&netops.Route{Destination: netops.DefaultPrefix})
+	if err != nil {
+		return netip.Prefix{}, nil, fmt.Errorf("failed to get routes on namespace %q: %w", ns.Path(), err)
+	}
+
+	for _, r := range routes {
+		if r.Device == iface && r.Destination.Bits() == 0 && r.Priority < math.MaxInt {
```

Comment:
I find it a bit odd to have the default route logic duplicated here. What do you think about modifying `GetDefaultRoutes` to optionally accept an interface name and apply the filtering inside? That way, this code could reuse `GetDefaultRoutes` ?

Reply:
For now, I'll keep it as is and avoid changing the interfaces since this code is experimental. It's also possible to provide another method as well.  When we work on adding support for other providers, I'll look at refactoring/redesigning it.